### DISCUSSION TITLE
Collection querying support

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#7068] [Design Proposals]
The idea is to allow querying and DML for collections roles which would make a `CollectionPersister` the root. Syntax suggestion for referring to `CollectionPersister` is `EntityName.collectionAttribute` and also through `fqn.EntityClass.collectionAttribute`.
I'm pretty sure that much of the custom SQL bits in Envers could be replaced with this new feature, given that I recently saw audited collections being implemented by a specially named dynamic entity.

**Comment by gavinking** [comment]
Funnily, enough, IIRC, we used to have something quite like this in much older versions of Hibernate, and it was removed at some stage (probably because the actual implementation sucked). Re-adding something like this, but doing it properly, is something that had already occurred to me so 👍

**Reply by sebersole** [reply]
Well, yes we used to have a feature to filter a collection *instance*.  AKA, load a `Person` and then filter its `person.addresses` by an HQL fragment.  I think (though could be wrong) that that is not what Christian is talking about.

Yes, that filtering feature was removed for a number of reasons - implementation was not great, "contract" was not well defined, kludgy syntax, etc.

**Comment by erikgollot** [comment]
Hi,

I had the following discussion with Christian https://discourse.hibernate.org/t/optimize-left-join-fetch/8280/16

My usecase is as follow:
A Person has some Cats and some Dogs.
Application A issue a first select of some persons, without cats and dogs.
Then Application A sends the persons with their data to Application B to filter the list. The role of Application B is to filter the Person list using the person data plus other information known by Application B.
Application B returns a list of Person’s Ids
Application A needs to load all Dogs of the corresponding Ids

Now imagine that the loaded persons are persons whose id is : 1,2,3,…30_000
Application B returns the ids : 1, 100, 2_000, 20_000, 25_000 up to 30_000

If I use @BatchSize(size=50) for example (whatever the size is)
If now I get the persons filtered by Application B and I do a person.getDogs() because I only need the Dogs of the filtrede Persons, I do not have any control on the persons for which Hibernate will load the Dogs.
Maybe, if I do a person.getDogs() for person 1, Hibernate will load persons for id = 1 up to 50. The problem is that I’m not interested by Persons 2 up to 50. The same will happen for person id=100,…and so on.
The problem with BatchSize is that the application does not have any control on the choice done by Hibernate

This is why I’ve asked for a syntax like “FROM Person p.id LEFT JOIN FETCH p.dogs” (or any other solution that allows me to retrieve only the id of a previously loaded object)…and the reason is that a Person has lot of data and returning them multiple time when I have to LEFT JOIN FETCH is not efficient.

Any comments ?

**Comment by dreab8** [comment]
https://hibernate.atlassian.net/browse/HHH-19406
